src/LinearSolvers/ColumnwiseGMRESSolver.jl

```diff
@@ -0,0 +1,177 @@
+module ColumnwiseGMRESSolver
+
+export StackGMRES, SingleColumnGMRES
+
+using ..DGmethods
+using ..DGmethods: BalanceLaw, DGModel
+using ..LinearSolvers
+const LS = LinearSolvers
+using ..MPIStateArrays: device, realview
+using LinearAlgebra
+using StaticArrays
+using GPUifyLoops
+
+abstract type AbstractColumnGMRESSolver{M} <: LS.AbstractIterativeLinearSolver end
+
+"""
+
+    StackGMRES(Q::AT; M, rtol, atol, nhorzelem) where AT
+
+This solver performs the matrix-free generalised minimal residual method for
+nodal columns residing in the same stack of elements. The stack-global
+convergence criterion requires convergence in all nodal columns.
+"""
+mutable struct StackGMRES{M, N, MP1, T, I, AT} <: AbstractColumnGMRESSolver{M}
+  krylov_basis::NTuple{MP1, AT}
+  "Hessenberg matrix"
+  H::Vector{Matrix{T}}
```

Comment:
Try a `Vector{S}` where `S<:SMatrix`, and same below, then in the main loop you can do `MMatrix(solver.H[1])` to get an `MMatrix` that you can mutate.